editor/code_editor.cpp

```diff
@@ -1942,6 +1890,44 @@ void CodeTextEditor::_toggle_scripts_pressed() {
 	update_toggle_scripts_button();
 }
 
+int CodeTextEditor::_get_affected_lines_from(int p_caret) {
+	if (text_editor->has_selection(p_caret)) {
+		return text_editor->get_selection_from_line(p_caret);
+	} else {
+		return text_editor->get_caret_line(p_caret);
+	}
+}
+
+int CodeTextEditor::_get_affected_lines_to(int p_caret) {
+	if (text_editor->has_selection(p_caret)) {
+		int line = text_editor->get_selection_to_line(p_caret);
+		// Don't affect a line with no selected characters
+		if (text_editor->get_selection_to_column(p_caret) == 0) {
+			line--;
+		}
+		return line;
+	} else {
+		return text_editor->get_caret_line(p_caret);
+	}
+}
```

Comment:
> There quite a lot here, gave a initial look over.
> 
> I do want to move these methods into `CodeEdit`, so if you feel up to it feel free to move them there and then we can take advantage of some unit tests.

Yes, this is what I've been thinking as well. I can try to do it. I have no experience in writing unit tests, but I'll write on contributor chat if I need help with those.

To be clear, do you think I should modify this pr to move them, or should we finish this first and then do the move? Do we have time to make it in 4.0 (since it would be an addition to the api) or wait for 4.1? Since rc phase is "just days away", I'd rather have more robust editor for 4.0 and nice additions to CodeEdit on next release.
Also, here's a list of operations I think should be moved into CodeEdit, do you agree?
* Delete lines
* Toggle comments
* Toggle bookmarks
* Toggle breakpoints (this needs a bit of extra logic from script editor to add breakpoints into debugger)
* Move lines up/down
* Duplicate selection (haven't touched to this in this pr, but also makes sense)